backend/chunker/chunker.go

```diff
@@ -0,0 +1,1412 @@
+
+	for _, chunk := range o.chunks {
+		if limit <= 0 {
+			break
+		}
+		count := chunk.Size()
+		if offset >= count {
+			offset -= count
+			continue
+		}
+		count -= offset
+		if limit < count {
+			count = limit
+		}
+
+		end := offset + count - 1
+		chunkOptions := append(openOptions, &fs.RangeOption{Start: offset, End: end})
+		rc, err := chunk.Open(ctx, chunkOptions...)
+		if err != nil {
+			_ = mrc.Close() // ignore error
+			return nil, err
+		}
+		mrc.readClosers = append(mrc.readClosers, rc)
+		readers = append(readers, rc)
+
+		offset = 0
+		limit -= count
+	}
+
+	mrc.multiReader = io.MultiReader(readers...)
+	return &mrc, nil
+}
+
+type multiReadCloser struct {
+	readClosers []io.ReadCloser
+	multiReader io.Reader
+}
+
+func (mrc *multiReadCloser) Read(p []byte) (n int, err error) {
+	return mrc.multiReader.Read(p)
+}
+
+func (mrc *multiReadCloser) Close() (err error) {
+	for _, rc := range mrc.readClosers {
+		chunkErr := rc.Close()
+		if err == nil {
+			err = chunkErr
+		}
+	}
+	return
+}
+
+// newDir returns a wrapped dir
+func (f *Fs) newDir(ctx context.Context, dir fs.Directory) fs.Directory {
+	newDir := fs.NewDirCopy(ctx, dir)
+	remote := dir.Remote()
+	newDir.SetRemote(remote)
+	return newDir
+}
+
+// UserInfo returns info about the connected user
+func (f *Fs) UserInfo(ctx context.Context) (map[string]string, error) {
+	do := f.Fs.Features().UserInfo
+	if do == nil {
+		return nil, fs.ErrorNotImplemented
+	}
+	return do(ctx)
+}
+
+// Disconnect the current user
+func (f *Fs) Disconnect(ctx context.Context) error {
+	do := f.Fs.Features().Disconnect
+	if do == nil {
+		return fs.ErrorNotImplemented
+	}
+	return do(ctx)
+}
+
+// ObjectInfo describes a wrapped fs.ObjectInfo for being the source
+//
+// This sums the size of object chunks
+type ObjectInfo struct {
+	fs.ObjectInfo
+	f         *Fs
+	newRemote string
+	newSize   int64
+	newTime   *time.Time
+}
+
+func (f *Fs) newInfo(src fs.ObjectInfo, newRemote string, newSize int64, newTime *time.Time) *ObjectInfo {
+	return &ObjectInfo{
+		ObjectInfo: src,
+		f:          f,
+		newRemote:  newRemote,
+		newSize:    newSize,
+		newTime:    newTime,
+	}
+}
+
+// Fs returns read only access to the Fs that this object is part of
+func (oi *ObjectInfo) Fs() fs.Info {
+	return oi.f
+}
+
+// Remote returns the remote path
+func (oi *ObjectInfo) Remote() string {
+	if oi.newRemote != "" {
+		return oi.newRemote
+	}
+	return oi.ObjectInfo.Remote()
+}
+
+// Size returns the size of the file
+func (oi *ObjectInfo) Size() int64 {
+	if oi.newSize != -1 {
+		return oi.newSize
+	}
+	return oi.ObjectInfo.Size()
+}
+
+// ModTime returns the modification time
+func (oi *ObjectInfo) ModTime(ctx context.Context) time.Time {
+	if oi.newTime != nil {
+		return *oi.newTime
+	}
+	return oi.ObjectInfo.ModTime(ctx)
+}
+
+// Hash returns the selected checksum of the file
+// If no checksum is available it returns ""
+func (oi *ObjectInfo) Hash(ctx context.Context, hash hash.Type) (string, error) {
```

Comment:
Could we put the Hash of the file into the metadata?

Reply:
See above